x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/update/update_gaps.ts

```diff
@@ -0,0 +1,219 @@
+          eventLogClient,
+          logger,
+          params: {
+            gapId: currentGap.internalFields._id,
+            ruleId,
+          },
+        });
+
+        if (!updatedGap) {
+          logger.warn(
+            `Gap ${currentGap.internalFields._id} not found during retry, it was removed by another process`
+          );
+          return;
+        }
+        currentGap = updatedGap;
+      }
+
+      const hasFailedBackfillTask = backfillSchedule?.some(
+        (scheduleItem) =>
+          scheduleItem.status === adHocRunStatus.ERROR ||
+          scheduleItem.status === adHocRunStatus.TIMEOUT
+      );
+
+      if (backfillSchedule && !hasFailedBackfillTask) {
+        updateGapFromSchedule({
+          gap: currentGap,
+          backfillSchedule,
+        });
+      }
+
+      if (hasFailedBackfillTask || !backfillSchedule || shouldRefetchAllBackfills) {
+        await calculateGapStateFromAllBackfills({
+          gap: currentGap,
+          savedObjectsRepository,
+          ruleId,
+          backfillClient,
+          actionsClient,
+        });
+      }
+
+      const esGap = currentGap.getEsObject();
+      const internalFields = currentGap.internalFields;
+
+      if (internalFields) {
+        await alertingEventLogger.updateGap({
+          internalFields,
+          gap: esGap,
+        });
+      }
+      return;
+    } catch (e) {
+      if (e.statusCode === CONFLICT_STATUS_CODE) {
+        retryCount++;
+        if (retryCount > MAX_RETRIES) {
+          logger.error(
+            `Failed to update gap: ${currentGap?.internalFields?._id} after ${MAX_RETRIES} retries due to conflicts`
+          );
+          throw e;
+        }
+        // Use exponential backoff with some randomness: 2s, 4s, 8s
+        const retryDelaySec = Math.min(Math.pow(2, retryCount), 30);
+        await delay(retryDelaySec * 1000 * Math.random());
+        continue;
+      }
+      throw e;
+    }
+  }
+};
+
+export const updateGaps = async (params: UpdateGapsParams) => {
+  const {
+    ruleId,
+    start,
+    end,
+    logger,
+    eventLogClient,
+    eventLogger,
+    backfillSchedule,
+    savedObjectsRepository,
+    shouldRefetchAllBackfills,
+    backfillClient,
+    actionsClient,
+  } = params;
+
+  if (!eventLogger) {
+    throw new Error('Event logger is required');
+  }
+
+  try {
+    const alertingEventLogger = new AlertingEventLogger(eventLogger);
+    let currentPage = 1;
+    const perPage = 10000;
+    let hasErrors = false;
+
+    while (true) {
+      const { data: gaps } = await findGaps({
+        eventLogClient,
+        logger,
+        params: {
+          ruleId,
+          start: start.toISOString(),
+          end: end.toISOString(),
+          page: currentPage,
+          perPage,
+          statuses: [gapStatus.PARTIALLY_FILLED, gapStatus.UNFILLED],
+        },
+      });
+
+      for (const gap of gaps) {
+        try {
+          await updateSingleGap(gap, {
+            backfillSchedule,
```

Comment:
it looks like we're using the same backfill schedule in each of these `updateSingleGap` calls, which then calls `updateGapFromSchedule` which goes through each schedule entry and converts the schedule entry into an interval. It might be better to do this processing of the schedule up front in the case where we have a lot of gaps to update, then we're not doing repeated processing of the same data.